File: services/kotlin/src/main/kotlin/com/corestack/service/CommentService.kt
package com.corestack.service

import com.corestack.repository.*
import io.github.oshai.kotlinlogging.KotlinLogging
import java.util.UUID

private val logger = KotlinLogging.logger {}

class CommentService(private val repo: CommentRepo) {

    suspend fun create(params: CreateCommentParams): Comment {
        if (params.taskId == UUID(0, 0)) throw InvalidInputError("invalid input: task_id is required")
        if (params.authorId.isBlank()) throw InvalidInputError("invalid input: author_id is required")
        if (params.content.isBlank()) throw InvalidInputError("invalid input: content is required")
        logger.debug { "creating comment task_id=${params.taskId} author_id=${params.authorId}" }
        return repo.create(params)
    }

    suspend fun list(params: ListCommentsParams): CommentList = repo.list(params)

    suspend fun delete(id: UUID) = repo.delete(id)
}

File: services/typescript/src/service/comment-service.ts
import { InvalidInputError } from "../repository/errors.js";
import {
  CommentRepo,
  type Comment,
  type CommentList,
  type CreateCommentParams,
  type ListCommentsParams,
} from "../repository/comment-repo.js";
import type { Logger } from "pino";

const EMPTY_UUID = "00000000-0000-0000-0000-000000000000";

export class CommentService {
  constructor(
    private repo: CommentRepo,
    private logger: Logger,
  ) {}

  async create(params: CreateCommentParams): Promise<Comment> {
    if (!params.taskId || params.taskId === EMPTY_UUID) {
      throw new InvalidInputError("invalid input: task_id is required");
    }
    if (!params.authorId) {
      throw new InvalidInputError("invalid input: author_id is required");
    }
    if (!params.content) {
      throw new InvalidInputError("invalid input: content is required");
    }
    this.logger.debug({ taskId: params.taskId, authorId: params.authorId }, "creating comment");
    return this.repo.create(params);
  }

  async list(params: ListCommentsParams): Promise<CommentList> {
    return this.repo.list(params);
  }

  async delete(id: string): Promise<void> {
    return this.repo.delete(id);
  }
}
